File: MetaBrain/observability/provider-metrics.ts
import type { ObservabilityEvent, ProviderMetrics } from "./types";

export function calculateProviderMetrics(provider_name: string, events: ObservabilityEvent[]): ProviderMetrics {
  const providerEvents = events.filter((event) => event.layer === "provider" && event.payload_summary.provider_name === provider_name);
  const requestCount = providerEvents.length;
  const timeoutCount = providerEvents.filter((event) => event.event_type.includes("timeout")).length;
  const fallbackCount = providerEvents.filter((event) => event.event_type.includes("fallback")).length;
  const errorCount = providerEvents.filter((event) => event.severity === "error" || event.severity === "critical").length;
  const latencies = providerEvents.map((event) => Number(event.payload_summary.latency_ms)).filter(Number.isFinite);
  return {
    provider_name,
    request_count: requestCount,
    timeout_count: timeoutCount,
    fallback_count: fallbackCount,
    latency_avg_ms: latencies.length ? Math.round(latencies.reduce((sum, value) => sum + value, 0) / latencies.length) : 0,
    degraded_mode_count: providerEvents.filter((event) => event.event_type.includes("degraded")).length,
    error_rate: requestCount ? Number((errorCount / requestCount).toFixed(3)) : 0,
  };
}

File: MB-Chat/observability_py/provider_metrics.py
from .types import ObservabilityEvent, ProviderMetrics


def calculate_provider_metrics(provider_name: str, events: list[ObservabilityEvent]) -> ProviderMetrics:
    scoped = [event for event in events if event.layer == "provider" and event.payload_summary.get("provider_name") == provider_name]
    latencies = [float(event.payload_summary["latency_ms"]) for event in scoped if isinstance(event.payload_summary.get("latency_ms"), (int, float))]
    errors = [event for event in scoped if event.severity in {"error", "critical"}]
    return ProviderMetrics(
        provider_name=provider_name,
        request_count=len(scoped),
        timeout_count=len([event for event in scoped if "timeout" in event.event_type]),
        fallback_count=len([event for event in scoped if "fallback" in event.event_type]),
        latency_avg_ms=round(sum(latencies) / len(latencies)) if latencies else 0,
        degraded_mode_count=len([event for event in scoped if "degraded" in event.event_type]),
        error_rate=round(len(errors) / len(scoped), 3) if scoped else 0.0,
    )
